core/src/consensus.rs

```diff
@@ -606,6 +703,203 @@ impl Tower {
+        let mut ancestors: HashSet<_> = self.stray_restored_slots.clone().into_iter().collect();
+        if let Some(last_voted_slot) = self.last_voted_slot() {
+            assert!(ancestors.contains(&last_voted_slot));
+            ancestors.remove(&last_voted_slot);
+        }
+        ancestors
+    }
+
+    // The tower root can be older/newer if the validator booted from a newer/older snapshot, so
+    // tower lockouts may need adjustment
+    pub fn adjust_lockouts_after_replay(
+        self,
+        replayed_root_slot: Slot,
+        slot_history: &SlotHistory,
+    ) -> Result<Self> {
+        info!(
+            "adjusting lockouts (after replay up to {}): {:?}",
+            replayed_root_slot,
+            self.voted_slots()
+        );
+
+        assert_eq!(slot_history.check(replayed_root_slot), Check::Found);
+        // reconcile_blockstore_roots_with_tower() should already have aligned these.
+        assert!(
+            self.root().is_none() || self.root().unwrap() <= replayed_root_slot,
+            format!(
+                "tower root: {:?} >= replayed root slot: {}",
+                self.root().unwrap(),
+                replayed_root_slot
+            )
+        );
+        assert!(
+            self.last_vote == Vote::default() && self.lockouts.votes.is_empty()
+                || self.last_vote != Vote::default() && !self.lockouts.votes.is_empty(),
+            format!(
+                "last vote: {:?} lockouts.votes: {:?}",
+                self.last_vote, self.lockouts.votes
+            )
+        );
+
+        // return immediately if votes are empty...
+        if self.lockouts.votes.is_empty() {
+            assert_eq!(self.root(), None);
+            return Ok(self);
+        }
+
+        let last_voted_slot = self.last_voted_slot().unwrap();
+        if slot_history.check(last_voted_slot) == Check::TooOld {
+            // We could try hard to anchor with other older votes, but opt to simplify the
+            // following logic
+            return Err(TowerError::TooOldTower(
+                last_voted_slot,
+                slot_history.oldest(),
+            ));
+        }
+
+        self.do_adjust_lockouts_after_replay(replayed_root_slot, slot_history)
+    }
+
+    fn do_adjust_lockouts_after_replay(
+        mut self,
+        replayed_root_slot: Slot,
+        slot_history: &SlotHistory,
+    ) -> Result<Self> {
+        // retained slots will be consisted only from divergent slots
+        let mut retain_flags_for_each_vote_in_reverse: Vec<_> =
+            Vec::with_capacity(self.lockouts.votes.len());
+
+        let mut still_in_future = true;
+        let mut past_outside_history = false;
+        let mut checked_slot = None;
+        let mut anchored_slot = None;
+
+        let mut root_checked = false;
+        let mut checked_slots = if let Some(root) = self.lockouts.root_slot {
+            root_checked = true;
+            vec![root]
```

Comment:
Extending [this logic](https://github.com/solana-labs/solana/pull/9902#issuecomment-627875209), I needed to include root for this anchoring process, because if all slots in tower except the root was on minor fork, [this error check incorrectly triggers](https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR856-R857).

This might sound really edge-case? Unfortunately, this isn't. This always occurs with single-leader clusters such as `./run.sh`.